components/proxy/plugins/configcat/configcat.go

```diff
@@ -0,0 +1,212 @@
+	m           sync.RWMutex
+
+	httpClient *http.Client
+	logger     *zap.Logger
+}
+
+// CaddyModule returns the Caddy module information.
+func (ConfigCat) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{
+		ID:  "http.handlers.gitpod_configcat",
+		New: func() caddy.Module { return new(ConfigCat) },
+	}
+}
+
+// ServeHTTP implements caddyhttp.MiddlewareHandler.
+func (c *ConfigCat) ServeHTTP(w http.ResponseWriter, r *http.Request, next caddyhttp.Handler) error {
+	if !pathRegex.MatchString(r.URL.Path) {
+		return next.ServeHTTP(w, r)
+	}
+	w.Header().Set("Content-Type", "application/json")
+	if c.sdkKey == "" {
+		w.Write(DefaultConfig)
+		return nil
+	}
+	etag := r.Header.Get("If-None-Match")
+	arr := strings.Split(r.URL.Path, "/")
+	configVersion := arr[len(arr)-1]
+	b, hash := c.getConfigCatJSON(configVersion)
+	if etag != "" && hash == etag {
+		w.WriteHeader(http.StatusNotModified)
+		return nil
+	}
+	if hash != "" {
+		w.Header().Set("ETag", hash)
+	}
+	w.Write(b)
+	return nil
+}
+
+func (c *ConfigCat) Provision(ctx caddy.Context) error {
+	c.logger = ctx.Logger(c)
+	c.configCache = make(map[string]configCache)
+
+	c.sdkKey = os.Getenv("CONFIGCAT_SDK_KEY")
+	if c.sdkKey == "" {
+		return nil
+	}
+
+	c.httpClient = &http.Client{
+		Timeout: 10 * time.Second,
+	}
+	c.baseUrl = os.Getenv("CONFIGCAT_BASE_URL")
+	if c.baseUrl == "" {
+		c.baseUrl = "https://cdn-global.configcat.com"
+	}
+	dur, err := time.ParseDuration(os.Getenv("CONFIGCAT_POLL_INTERVAL"))
+	if err != nil {
+		c.pollInterval = time.Minute
+		c.logger.Warn("cannot parse poll interval of configcat, default to 1m")
+	} else {
+		c.pollInterval = dur
+	}
+
+	// poll config
+	go func() {
+		for range time.Tick(c.pollInterval) {
+			for version := range c.configCache {
+				data, err := c.fetchConfigCatConfig(version)
+				if err != nil {
+					continue
+				}
+				c.updateCache(version, data)
+			}
+		}
+	}()
+	return nil
+}
+
+func (c *ConfigCat) updateCache(configVersion string, data []byte) (hash string) {
+	sum := md5.Sum(data)
+	hash = hex.EncodeToString(sum[:])
+	c.m.Lock()
+	c.configCache[configVersion] = configCache{
+		data: data,
+		hash: hash,
+	}
+	c.m.Unlock()
+	return hash
+}
+
+func (c *ConfigCat) getConfigCatJSON(configVersion string) ([]byte, string) {
+	c.m.RLock()
+	cache, ok := c.configCache[configVersion]
+	c.m.RUnlock()
+	if ok {
+		return cache.data, cache.hash
+	}
+	t, err := c.fetchConfigCatConfig(configVersion)
+	if err != nil {
+		return DefaultConfig, ""
+	}
+	hash := c.updateCache(configVersion, t)
+	return t, hash
+}
+
+var sg = &singleflight.Group{}
+
+// fetchConfigCatConfig with different config version. i.e. config_v5.json
+func (c *ConfigCat) fetchConfigCatConfig(configVersion string) ([]byte, error) {
+	b, err, _ := sg.Do(fmt.Sprintf("fetch_%s", configVersion), func() (interface{}, error) {
+		url := fmt.Sprintf("%s/configuration-files/%s/%s", c.baseUrl, c.sdkKey, configVersion)
+		resp, err := c.httpClient.Get(url)
```

Comment:
Send the "If-None-Match" header to avoid fetching config.json if it hasn't change, check https://github.com/configcat/go-sdk/blob/master/v7/config_fetcher.go#L300-L344

Reply:
`If-None-Match` located at request [header](https://github.com/configcat/go-sdk/blob/master/v7/config_fetcher.go#L311) for server to check with etag (from previous config fetch), client will check if it's 304, if so, [use previous one](https://github.com/configcat/go-sdk/blob/eaa908ee61da13a1786a9dcda18de41c53d6be66/v7/config_fetcher.go#L320-L323)

Reply:
I see () 😅

Reply:
I pushed a commit to request with etag @jeanp413